### PR TITLE
Add support for skipped columns

### PR DESCRIPTION
**Pull request by matiasfrndz**
Based on the feedback from @optilude and @rlotufo for #24 and #26 I refactored the code that is supposed to add support for skipped columns. I have refactored `cycletime.py` so that it calculates the correct timestamp for skipped columns, mimicking the behaviour of ActionableAgile Analytics. This makes calculations in subsequent calculators much easier, because they do no't have to handle these special cases anymore.

I also refactored a small part of `config.py` to perform some basic checks on the `Workflow` settings:

- The `Backlog column` has to be the column immediately preceding the `Committed column`
- There must at least one column after the `Committed column`, that is the `Done column`

All unit tests run without issues and I have already tested with different Jira setups I have in production.

### CODE REVIEW

File: jira_agile_metrics/calculators/ageingwip.py [hunk 1]
```diff
@@ -24,19 +24,9 @@ def run(self, today=None):
...
-        
-        if start_column not in cycle_names:
-            logger.error("Committed column %s does not exist", start_column)
-            return None
-        if end_column not in cycle_names:
-            logger.error("Final column %s does not exist", end_column)
-            return None
-        if done_column not in cycle_names:
-            logger.error("Done column %s does not exist", done_column)
-            return None
```
Comment (optilude) [inline jira_agile_metrics/calculators/ageingwip.py:39]: Why remove these checks? They are useful for alerting the user if their configuration is internally inconsistent.

File: jira_agile_metrics/calculators/ageingwip.py [hunk 2]
```diff
@@ -24,19 +24,9 @@ def run(self, today=None):
...
+        last_active_column = cycle_names[cycle_names.index(done_column)-1]
```
Comment (optilude) [inline jira_agile_metrics/calculators/ageingwip.py:29]: The column described by the `end_column` config parameter is meant to allow the user to set this as a column that might not be the second-to-last-one (although it defaults to that, see https://github.com/DeloitteDigitalUK/jira-agile-metrics/blob/master/jira_agile_metrics/config.py#L477).

This is helpful because it allows you to define a "wide" cycle (e.g. in a file extended by multiple configurations) but also run certain analysis over a subset (we have a shared JIRA workflow that goes from very early analysis to live, but some of our analytics run on a subset of this workflow, e.g. a "dev team" flow).

Is there a reason it has to be hardcoded like this?

File: jira_agile_metrics/calculators/cycletime.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ def run(self, now=None):
             self.query_manager,
             self.settings['cycle'],
             self.settings['attributes'],
-            self.settings['backlog_column'],
+            self.settings['committed_column'],
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:46]: Can you explain why we are changing this? Does it alter the meaning of the configuration options? I'm very worried about breaking backwards compatibility with existing configurations in the wild.

File: jira_agile_metrics/calculators/cycletime.py [hunk 4]
```diff
@@ -95,16 +95,14 @@ def calculate_cycle_times(
         now = datetime.datetime.utcnow()
 
     cycle_names = [s['name'] for s in cycle]
-    accepted_steps = set(s['name'] for s in cycle if s['type'] == StatusTypes.accepted)
-    completed_steps = set(s['name'] for s in cycle if s['type'] == StatusTypes.complete)
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:99]: I kind of see why this makes sense. I think originally the idea was that we'd use the `StatusTypes` enum to mark columns as representing backlog, committed, active, done, etc. states. But in reality you're still reliant on applying the analysis to a sequential set of columns from start to finish.

Not really for this PR but I wonder if we should just get rid of the `type` option against cycles altogether, i.e. is it really doing anything?

File: jira_agile_metrics/calculators/cycletime.py [hunk 5]
```diff
@@ -197,8 +195,8 @@ def calculate_cycle_times(
                         if impediment_start is None:
                             logger.warning("Issue %s had impediment flag cleared before being set. This should not happen.", issue.key)
                             continue
-                        
-                        if impediment_start_status not in (backlog_column, done_column):
+
+                        if impediment_start_status in active_columns:
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:199]: So yeah - these are nicer and less confusing with the `active_columns` pattern

File: jira_agile_metrics/calculators/cycletime.py [hunk 6]
```diff
@@ -237,25 +235,28 @@ def calculate_cycle_times(
...
+            # calculate cycle time
 
             previous_timestamp = None
-            accepted_timestamp = None
-            completed_timestamp = None
+            committed_timestamp = None
+            done_timestamp = None
 
-            for cycle_name in cycle_names:
+            for cycle_name in reversed(cycle_names):
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:245]: Just so that I can make sure I am understanding this correctly, could you explain the thinking behind doing this check in reverse order and what the implications are? Does this fix your bug?

File: jira_agile_metrics/calculators/cycletime.py [hunk 7]
```diff
@@ -237,25 +235,28 @@ def calculate_cycle_times(
...
                 if item[cycle_name] is not None:
                     previous_timestamp = item[cycle_name]
 
-                    if accepted_timestamp is None and previous_timestamp is not None and cycle_name in accepted_steps:
-                        accepted_timestamp = previous_timestamp
-                    if completed_timestamp is None and previous_timestamp is not None and cycle_name in completed_steps:
-                        completed_timestamp = previous_timestamp
+                if previous_timestamp is not None:
+                    item[cycle_name] = previous_timestamp
+                    if cycle_name == done_column:
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:251]: If I'm thinking about this correctly, one implication of this change (basically, using an `==` instead of an `in` check) and the change to how we identify the "start" and "end" column(s) is that before there could in theory be more than one "accepted" (my language) / "committed' (your language - which is better, I agree) column, and now there can be only one. And the same for "completed"/"done". Correct?

I think that's probably fine. It relates to the point above about trying to design for there potentially being multiple columns with `type==StatusTypes.accepted` (and `complete`), but I'm not really sure that's a valid use case. I think it's an overhang from before I implemented mapping of multiple JIRA statuses to a single logical workflow step in this tool.

It does - and I think you said this before - mean that if you have multiple things after "done" *and* the ticket never touches the (single) "done" state but skips straight to one of the later ones, then it'll appear to not have been completed. Again, that's a slight change of behaviour but I'm not sure it massively matters as it seems unlikely people will have this happen.

Do you agree?

File: jira_agile_metrics/calculators/cycletime.py [hunk 8]
```diff
@@ -237,25 +235,28 @@ def calculate_cycle_times(
...
+            for cycle_name in reversed(cycle_names):
                 if item[cycle_name] is not None:
                     previous_timestamp = item[cycle_name]
 
-                    if accepted_timestamp is None and previous_timestamp is not None and cycle_name in accepted_steps:
-                        accepted_timestamp = previous_timestamp
-                    if completed_timestamp is None and previous_timestamp is not None and cycle_name in completed_steps:
-                        completed_timestamp = previous_timestamp
+                if previous_timestamp is not None:
+                    item[cycle_name] = previous_timestamp
```
Comment (optilude) [inline jira_agile_metrics/calculators/cycletime.py:250]: What is this line doing?

I am probably being a bit slow but I'm confused by the idea that we're going backwards through the cycle, and then setting the timestamp at the current cycle step to the timestamp of the "previous" in the iteration (which presumably is the *next* in the cycle)?

Can you explain the logic?

File: jira_agile_metrics/calculators/impediments.py [hunk 9]
```diff
@@ -32,21 +32,23 @@ def run(self):
         ):
             logger.debug("Not calculating impediments data as no output files specified")
             return None
-        
-        backlog_column = self.settings['backlog_column']
-        done_column = self.settings['done_column']
```
Comment (optilude) [inline jira_agile_metrics/calculators/impediments.py:38]: Again, could this change break configurations that rely on using these configuration options?

File: jira_agile_metrics/calculators/impediments_test.py [hunk 10]
```diff
@@ -155,13 +155,13 @@ def test_calculate_impediments(query_manager, settings, cycle_time_results):
 def test_different_backlog_column(query_manager, settings, cycle_time_results):
     settings = extend_dict(settings, {
         'backlog_column': 'Committed',
+        'committed_column': 'Build',
```
Comment (optilude) [inline jira_agile_metrics/calculators/impediments_test.py:158]: Why did you change this test?

Feels like this might mean changing previous behaviour?

If not and you just want to test a different scenario, could we leave this test and add a new one?

File: jira_agile_metrics/calculators/impediments_test.py [hunk 11]
```diff
@@ -175,6 +175,5 @@ def test_different_done_column(query_manager, settings, cycle_time_results):
 
     assert data.to_dict('records') == [
         {'key': 'A-2', 'status': 'Committed', 'flag': 'Impediment',     'start': _ts('2018-01-10'), 'end': _ts('2018-01-12')},
-        {'key': 'A-3', 'status': 'Done',      'flag': 'Impediment',     'start': _ts('2018-01-07'), 'end': _ts('2018-01-10')},
```
Comment (optilude) [inline jira_agile_metrics/calculators/impediments_test.py:178]: Again, it makes me nervous that you've changed the expected test output without changing anything else here. Can you explain your thinking?

File: jira_agile_metrics/calculators/progressreport.py [hunk 12]
```diff
@@ -443,11 +437,14 @@ def calculate_team_throughput(
     frequency
 ):
 
+    backlog_column_index = [s['name'] for s in cycle].index(backlog_column)
+    committed_column = cycle[backlog_column_index + 1]['name']  # config parser ensures that `backlog` comes immediately before `committed`
```
Comment (optilude) [inline jira_agile_metrics/calculators/progressreport.py:441]: This looks weird. Why did you make this change? Don't have a configuration parameter to identify the committed column? And also, I'm not sure the config parser does (intentionally?) ensure it's always "backlog column+1", that's just the default if the committed-column isn't specified.

File: jira_agile_metrics/calculators/progressreport.py [hunk 13]
```diff
@@ -506,13 +503,13 @@ def update_story_counts(
     done_column
 ):
     backlog_column_index = [s['name'] for s in cycle].index(backlog_column)
-    started_column = cycle[backlog_column_index + 1]['name']  # config parser ensures there is at least one column after backlog
-    
+    committed_column = cycle[backlog_column_index + 1]['name']  # config parser ensures that `backlog` comes immediately before `committed`
```
Comment (optilude) [inline jira_agile_metrics/calculators/progressreport.py:506]: ... or maybe I made that assumption. :) Is it right?

File: jira_agile_metrics/calculators/waste_test.py [hunk 14]
```diff
@@ -161,13 +161,12 @@ def test_query(jira, settings):
     assert data.to_dict('records') == [
         {'key': 'A-1', 'last_status': 'Test',      'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
         {'key': 'A-2', 'last_status': 'Committed', 'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-07 02:02:02')},
-        {'key': 'A-6', 'last_status': 'foobar',    'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
-        {'key': 'A-7', 'last_status': None,        'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
```
Comment (optilude) [inline jira_agile_metrics/calculators/waste_test.py:165]: Again, you've changed a test here and it's not obvious if this is just accepting a regression or fixing a genuinely invalid test. Can you elaborate?

File: jira_agile_metrics/calculators/waste_test.py [hunk 15]
```diff
@@ -178,9 +177,6 @@ def test_different_backlog_column(jira, settings):
 
     assert data.to_dict('records') == [
         {'key': 'A-1', 'last_status': 'Test',      'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
-        {'key': 'A-4', 'last_status': 'Backlog',   'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-09 02:02:02')},
-        {'key': 'A-6', 'last_status': 'foobar',    'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
-        {'key': 'A-7', 'last_status': None,        'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
```
Comment (optilude) [inline jira_agile_metrics/calculators/waste_test.py:183]: ... and here

File: jira_agile_metrics/calculators/waste_test.py [hunk 16]
```diff
@@ -196,7 +192,4 @@ def test_different_done_column(jira, settings):
 
     assert data.to_dict('records') == [
         {'key': 'A-2', 'last_status': 'Committed', 'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-07 02:02:02')},
-        {'key': 'A-3', 'last_status': 'Done',      'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-08 02:02:02')},
-        {'key': 'A-6', 'last_status': 'foobar',    'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
-        {'key': 'A-7', 'last_status': None,        'resolution': 'Withdrawn', 'withdrawn_date': Timestamp('2018-01-06 02:02:02')},
     ]
```
Comment (optilude) [inline jira_agile_metrics/calculators/waste_test.py:195]: ... and here

File: jira_agile_metrics/config.py [hunk 17]
```diff
@@ -390,7 +389,6 @@ def config_to_options(data, cwd=None, extended=False):
         for key in [
             'backlog_column',
             'committed_column',
-            'final_column',
```
Comment (optilude) [inline jira_agile_metrics/config.py:393]: Will this not regress existing configurations that rely on this parameter?

File: jira_agile_metrics/config_test.py [hunk 18]
```diff
@@ -40,9 +40,8 @@ def test_config_to_options_minimal():
 
     assert options['settings']['backlog_column'] == 'Backlog'
     assert options['settings']['committed_column'] == 'In progress'
-    assert options['settings']['final_column'] == 'In progress'
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:43]: And again - changes a test that would have passed previously.

File: jira_agile_metrics/config_test.py [hunk 19]
```diff
@@ -89,17 +88,16 @@ def test_config_to_options_maximal():
 
     Backlog column: Backlog
     Committed column: Committed
-    Final column: Test
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:92]: And again - changes a test that would have passed previously.

File: jira_agile_metrics/config_test.py [hunk 20]
```diff
@@ -236,16 +234,14 @@ def test_config_to_options_maximal():
...
-            {'name': 'Build', 'statuses': ['Build'], 'type': 'accepted'},
-            {'name': 'Test',
-                'statuses': ['Code review', 'QA'],
-                'type': 'accepted'},
-            {'name': 'Done', 'statuses': ['Done'], 'type': 'complete'}
+            {'name': 'Backlog', 'statuses': ['Backlog']},
+            {'name': 'Committed', 'statuses': ['Next']},
+            {'name': 'Build', 'statuses': ['Build']},
+            {'name': 'Test', 'statuses': ['Code review', 'QA']},
+            {'name': 'Done', 'statuses': ['Done']}
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:244]: Think this is OK - I can see you've taken out the `StatusTypes` concept which as per above is probably not all that useful. We should make sure we remove the enum as well if it's unused (and if it is used - why?)

File: jira_agile_metrics/config_test.py [hunk 21]
```diff
@@ -256,23 +252,22 @@ def test_config_to_options_maximal():
         'queries': [{'jql': '(filter=123)', 'value': 'Team 1'},
                     {'jql': '(filter=124)', 'value': 'Team 2'}],
         'query_attribute': 'Team',
-                
+
         'backlog_column': 'Backlog',
         'committed_column': 'Committed',
-        'final_column': 'Test',
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:262]: Again - changes a test that would have passed previously.

File: jira_agile_metrics/config_test.py [hunk 22]
```diff
@@ -450,8 +445,7 @@ def test_config_to_options_extends():
         - 0.2
 
     Backlog column: Backlog
-    Committed column: Committed
-    Final column: Test
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:454]: Again - changes a test that would have passed previously.

File: jira_agile_metrics/config_test.py [hunk 23]
```diff
@@ -482,16 +476,15 @@ def test_config_to_options_extends():
 
         # overridden
         assert options['connection']['domain'] == 'https://bar.com'
-        
+
         # from extended base
         assert options['settings']['backlog_column'] == 'Backlog'
-        assert options['settings']['committed_column'] == 'Committed'
-        assert options['settings']['final_column'] == 'Test'
+        assert options['settings']['committed_column'] == 'In progress'
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:482]: Again - changes a test that would have passed previously. This one makes me more nervous because it looks like it's genuinely picking a different column as the "Committed" one without changing the test setup, which surely means a regression for existing configurations?

File: jira_agile_metrics/config_test.py [hunk 24]
```diff
@@ -523,7 +516,6 @@ def test_config_to_options_extends_blocked_if_no_explicit_working_directory():
 
     Backlog column: Backlog
     Committed column: Committed
-    Final column: Test
```
Comment (optilude) [inline jira_agile_metrics/config_test.py:526]: Again - changes a test that would have passed previously.

File: jira_agile_metrics/conftest.py [hunk 25]
```diff
@@ -83,18 +83,17 @@ def minimal_settings():
...
-            {'name': 'Backlog',   'statuses': ['Backlog'],           'type': 'backlog'},
-            {'name': 'Committed', 'statuses': ['Next'],              'type': 'accepted'},
-            {'name': 'Build',     'statuses': ['Build'],             'type': 'accepted'},
-            {'name': 'Test',      'statuses': ['Code review', 'QA'], 'type': 'accepted'},
-            {'name': 'Done',      'statuses': ['Done'],              'type': 'complete'}
+            {'name': 'Backlog',   'statuses': ['Backlog']},
+            {'name': 'Committed', 'statuses': ['Next']},
+            {'name': 'Build',     'statuses': ['Build']},
+            {'name': 'Test',      'statuses': ['Code review', 'QA']},
+            {'name': 'Done',      'statuses': ['Done']}
```
Comment (optilude) [inline jira_agile_metrics/conftest.py:90]: These are fine as per above I think

File: jira_agile_metrics/conftest.py [hunk 26]
```diff
@@ -83,18 +83,17 @@ def minimal_settings():
...
         ],
         'query_attribute': None,
         'queries': [{'jql': '(filter=123)', 'value': None}],
-        
+
         'backlog_column': 'Backlog',
         'committed_column': 'Committed',
-        'final_column': 'Test',
```
Comment (optilude) [inline jira_agile_metrics/conftest.py:97]: Again - changes a test that would have passed previously.

**Comment by optilude** [comment]
@matiasfrndz Thank you so much for this. I really appreciate the effort you've put into this and I think the changes are positive. And sorry it's taken me a little while to get around to looking at it.

However, there are a few tests that appear to have been changed which makes me nervous about regressions (I have teams relying on this tool for client work daily...), and there are a couple of areas where I'd love to just understand your thinking a bit better.

Have added a lot of detailed comments in the changelog. Do you mind having a look?

**Comment by matiasfrndz** [comment]
@optilude Thank you very much for reviewing my pull request so thoroughly, I highly appreciate that. Sorry for the delay of my response. Before I answer each of your questions I think it's helpful to describe my mental model. This is most easily done with a picture:
2020-05-30_jira-agile-metrics

* A Kanban board is used to describe and visualise a `Workflow` consisting of different stages visualised as columns
* When I a look at flow metrics, I'm interested to analyse them from a defined starting point to a defined end point
* I see the starting point as the moment an issue enters the `Committed column`, that is when it crosses the «commitment line»
* I see the end point as the moment when an issue enters the `Done column`
* I believe that all `Calculators` except `burnup`, `forecast`, `progressreport` require only these two settings, `Committed column` and `Done column`, to correctly perform calculations
* The following `Calculators` require an additional setting `Backlog column` to perform calculations: `burnup`, `forecast`, `progressreport`

Does that make sense?

Based on that I think that the attribute `StatusType` is not required. Also `Final column` doesn't seem to be necessary. Thus I decided to remove an occurrences of them in code. I totally understand that this makes you nervous, especially because I changed some tests. I you prefer, I can leave the tests as they where, however `Final column`doesn't have any effect whatsoever.

Now I'll answer each of the questions.

File: jira_agile_metrics/calculators/ageingwip.py (hunk 1 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/ageingwip.py:39]: These checks are done in `config.py: config_to_options` and I'd argue that it is more intuitive and robust to do such sanity checks once, and once only.  (See the [DRY principle](https://matiasfrndz.ch/2017/02/Dont-Repeat-Yourself-DRY)).

File: jira_agile_metrics/calculators/ageingwip.py (hunk 2 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/ageingwip.py:29]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I'd argue, that one can achieve the same effect with `Committed column` and `Done column`, that is by leaving the `Workflow` the same and just chaning `Committed column` and `Done column`.

File: jira_agile_metrics/calculators/cycletime.py (hunk 3 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/cycletime.py:46]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that cycle time calculations require `Committed column` and `Done column` only.

File: jira_agile_metrics/calculators/cycletime.py (hunk 4 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/cycletime.py:99]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that cycle time calculations require `Committed column` and `Done column` only.
Additional attributes would only be required if we wanted to ignore intermediate workflow steps (e.g. waiting columns) in cycle time calculation.

File: jira_agile_metrics/calculators/cycletime.py (hunk 6 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/cycletime.py:245]: When a column gets skipped, I use the next timestamp to fill in in skipped columns. For example if I have a workflow `A`, `B`, `C` and an issue goes from `A` directly to `C`, the I use the same timestamp from `C` for `B`. This is semantically correct and makes calculations much simpler. Semantically it means that an issue spent zero seconds in `B` before it moves on to `C`. ActionableAgile Analytics does it the same way.

File: jira_agile_metrics/calculators/cycletime.py (hunk 7 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/cycletime.py:251]: See the picture in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642, all workflow stages starting from `Committed column` until and excluding `Done column` are `active_columns`. Workflow stages after `Done column` can be ignored.

File: jira_agile_metrics/calculators/cycletime.py (hunk 8 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/cycletime.py:250]: Imagine workflow `A`, `B`, `C` and were going backwards starting at `C`. If `C` has a timestamp and `B` doesn't because the issue went straight from `A` to `C`, then I can use the timestamp from `C` for `B` too. This is semantically correct, because it means that the issue spent zero seconds in `B`, and it makes calculations easier. This is equal to the way ActionableAgile Analytics handels cases with skipped columns.

File: jira_agile_metrics/calculators/progressreport.py (hunk 13 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/progressreport.py:506]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 i think that the timer for cycle time starts ticking when an issue enters `Committed column`.  It was never explicitly stated that the `Backlog column` should immediately precede the `Committed column`. That's the reason I added an explicit check in `config.py`.

File: jira_agile_metrics/config.py (hunk 17 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config.py:393]: `final_column` was only used once in `ageingwip.py`, however, I changed the code to use `done_column` instead for the reasons described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642. I'm confident to say that this won't break anything.

File: jira_agile_metrics/config_test.py (hunk 19 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:92]: `Final column` is not used in the code, therefore it would be okay to leave that there. However, I think it's confusing to have code or config that isn't used.

File: jira_agile_metrics/config_test.py (hunk 20 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:244]: Yes, as described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think `StatusTypes` in not needed.

File: jira_agile_metrics/config_test.py (hunk 21 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:262]: I would be okay to leave it there. However, it's not needed and would be ignored. I find it confusing to have «dead» config in tests. Maybe it makes sense to add a comment noting that this is a legacy configuration parameter.

File: jira_agile_metrics/conftest.py (hunk 26 above)
Comment (matiasfrndz) [reply jira_agile_metrics/conftest.py:97]: It would be safe to leave that there because it's not used and would be ignored.

File: jira_agile_metrics/config_test.py (hunk 24 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:526]: It would be safe to leave that there because it's not used and would be ignored.

File: jira_agile_metrics/calculators/impediments_test.py (hunk 10 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/impediments_test.py:158]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that `impediments.py` should only consider `active_columns`, i.e. from `committed_column` up to and excluding `done_column`.

File: jira_agile_metrics/calculators/impediments_test.py (hunk 11 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/impediments_test.py:178]: This test is/was done based on the following config:
```
        'cycle': [
            {'name': 'Backlog',   'statuses': ['Backlog'],           'type': 'backlog'},
            {'name': 'Committed', 'statuses': ['Next'],              'type': 'accepted'},
            {'name': 'Build',     'statuses': ['Build'],             'type': 'accepted'},
            {'name': 'Test',      'statuses': ['Code review', 'QA'], 'type': 'accepted'},
            {'name': 'Done',      'statuses': ['Done'],              'type': 'complete'}
        ],
```
If `committed_column` is `Committed` and `done_column` is `Build`, then only the block should be reported:
```
{'start': date(2018, 1, 10), 'end': date(2018, 1, 12), 'status': 'Committed', 'flag': 'Impediment'},  # included
```

File: jira_agile_metrics/calculators/progressreport.py (hunk 12 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/progressreport.py:441]: See https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#discussion_r432833831 the implicit assumption was the following:
```
started_column = cycle[backlog_column_index + 1]['name']  # config parser ensures there is at least one column after backlog
```
I made this explicit in `config.py` ensuring that `backlog_column` is immediately preceding `committed_column`. These implicit assumptions confused me a lot when refactoring the code.

File: jira_agile_metrics/calculators/waste_test.py (hunk 14 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/waste_test.py:165]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that `waste.py` should only consider `active_columns` starting from `committed_column` up to and excluding `done_column`. Columns before `committed_columns` as well as after and including `done_column` should be ignored.

File: jira_agile_metrics/calculators/waste_test.py (hunk 15 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/waste_test.py:183]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that `waste.py` should only consider `active_columns` starting from `committed_column` up to and excluding `done_column`. Columns before `committed_columns` as well as after and including `done_column` should be ignored.

File: jira_agile_metrics/calculators/waste_test.py (hunk 16 above)
Comment (matiasfrndz) [reply jira_agile_metrics/calculators/waste_test.py:195]: As described in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#issuecomment-636308642 I think that `waste.py` should only consider `active_columns` starting from `committed_column` up to and excluding `done_column`. Columns before `committed_columns` as well as after and including `done_column` should be ignored.

File: jira_agile_metrics/config_test.py (hunk 18 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:43]: `final_column` is not used anymore, therefore I would find it confusing to have a test for it.

File: jira_agile_metrics/config_test.py (hunk 22 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:454]: This didn't make sense before:
```
Workflow:
    Backlog: Backlog
    In progress: Build
    Done: Done
```
and
```
    Backlog column: Backlog
    Committed column: Committed
    Final column: Test
    Done column: Done
```
This is wrong in two ways:
1. There is no workflow stage called `Committed`. When setting `Committed column: Committed` the sanity check should have thrown an exception.
2. As for me, the assumption in the assertion is wrong.

With the changes and new sanity checks in `config.py`, this config will trigger an exception.

File: jira_agile_metrics/config_test.py (hunk 23 above)
Comment (matiasfrndz) [reply jira_agile_metrics/config_test.py:482]: See https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/27#discussion_r432839745.

**Comment by matiasfrndz** [comment]
I now answered all your questions above. I hope they are helpful. Please don't hesitate asking if my provided answers are not clear or if you think that I have made wrong assumption.

**Comment by optilude** [comment]
Sorry it’s taken ages to reviews these and thank you both for your contributions and the detailed discussion!

**Comment by matiasfrndz** [comment]
Thank you @optilude , I'm happy you found my pull request and comments helpful.
